torch/testing/_internal/autograd_function_db.py

```diff
@@ -84,11 +89,19 @@ def backward(ctx, grad_output):
             gy = NumpyMul.apply(grad_output, x)
         return gx, gy
 
+    @staticmethod
+    def vmap(info, in_dims, x, y):
+        x_bdim, y_bdim = in_dims
+        x = x.movedim(x_bdim, -1) if x_bdim is not None else x.unsqueeze(-1)
+        y = y.movedim(y_bdim, -1) if y_bdim is not None else y.unsqueeze(-1)
+        result = NumpyMul.apply(x, y)
+        result = result.movedim(-1, 0)
+        return result, 0
 
 def sample_inputs_numpy_mul(opinfo, device, dtype, requires_grad, **kwargs):
     make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
     # Broadcasting
-    yield SampleInput(make_arg(2, low=0.9, high=2), args=(make_arg(3, 2, low=0.9, high=2),))
+    yield SampleInput(make_arg(4, low=0.9, high=2), args=(make_arg(3, 4, low=0.9, high=2),))
```

Comment:
sizes adjusted because we add extra dimensions of size 2 in vmap testing, so it is easier to debug when the sizes (before being pass through the vmap tests) don't have dimensions of size 2.

Reply:
As a vmap noob who hasn't debugged many batching rules, I don't understand this lol, could you elaborate a little?

Reply:
Yup!

Let's say I incorrectly wrote the mul batching rule as:
```py
def vmap(info, in_dims, x, y):
  return x * y
```

and then we tried:
```py
x = torch.randn(2, 2)
y = torch.randn(2, 2, 2)
z = vmap(NumpyMul.apply, in_dims(0, 0))(x, y)
```
This would succeed and also be wrong -- we would end up with z = x * y, but the result should be z = x.unsqueeze(1) * y (we need to line up dimension 0 of x with dimension 0 of y). The test runner would fail with something like "output values are incorrect".

If the dimensions being vmapped over for x and y have a unique shape from the rest of the dimensions, then we get a loud error:

```py
x = torch.randn(2, 4)
y = torch.randn(2, 3, 4)
z = vmap(NumpyMul.apply, in_dims(0, 0))(x, y)
# Error: can't broadcast Tensor of shape [2, 4] with Tensor of shape [2, 3, 4]
```
and with this error we can tell that we're doing something wrong (we forgot to change `x` into having shape [2, 1, 4] before multiplying it with y)

Now, how does this apply to the samples in the OpInfo?
- vmap testing takes the samples in the OpInfo and expands them to include an additional dimension of size 2.
- Then it just does vmap(op)(*expanded_inputs).
- Ideally the original samples in the OpInfo do not have dimensions of size 2 so we get a loud shape error instead of a "values are incorrect" error -- the shape error tells us we misaligned dimensions

Reply:
Makes sense, thanks!